src/Network/Request.php

```diff
@@ -944,22 +944,175 @@ public function here($base = true)
     }
 
     /**
+     * Normalize a header name into the SERVER version.
+     *
+     * @param string $name The header name.
+     * @return string The normalized header name.
+     */
+    protected function normalizeHeaderName($name)
+    {
+        $name = str_replace('-', '_', strtoupper($name));
+        if (!in_array($name, ['CONTENT_LENGTH', 'CONTENT_TYPE'])) {
+            $name = 'HTTP_' . $name;
+        }
+
+        return $name;
+    }
+
+    /**
      * Read an HTTP header from the Request information.
      *
+     * If the header is not defined in the request, this method
+     * will fallback to reading data from $_SERVER and $_ENV.
+     * This fallback behavior is deprecated, and will be removed in 4.0.0
+     *
      * @param string $name Name of the header you want.
      * @return string|null Either null on no header being set or the value of the header.
+     * @deprecated 4.0.0 The automatic fallback to env() will be removed in 4.0.0
      */
     public function header($name)
     {
-        $name = str_replace('-', '_', $name);
-        if (!in_array(strtoupper($name), ['CONTENT_LENGTH', 'CONTENT_TYPE'])) {
-            $name = 'HTTP_' . $name;
-        }
+        $name = $this->normalizeHeaderName($name);
 
         return $this->env($name);
     }
 
     /**
+     * Get all headers in the request.
+     *
+     * Returns an associative array where the header names are
+     * the keys and the values are a list of header values.
+     *
+     * While header names are not case-sensitive, getHeaders() will normalize
+     * the headers.
+     *
+     * @return array An associative array of headers and their values.
+     */
+    public function getHeaders()
+    {
+        $headers = [];
+        foreach ($this->_environment as $key => $value) {
+            $name = null;
+            if (strpos($key, 'HTTP_') === 0) {
+                $name = substr($key, 5);
+            }
+            if (strpos($key, 'CONTENT_') === 0) {
+                $name = $key;
+            }
+            if ($name !== null) {
+                $name = strtr(strtolower($name), '_', ' ');
+                $name = strtr(ucwords($name), ' ', '-');
+                $headers[$name] = (array)$value;
+            }
+        }
+
+        return $headers;
+    }
+
+    /**
+     * Check if a header exists in the request.
+     *
+     * @param string $name The header you want to get (case-insensitive)
+     * @return bool Whether or not the header is defined.
+     */
+    public function hasHeader($name)
+    {
+        $name = $this->normalizeHeaderName($name);
+
+        return array_key_exists($name, $this->_environment);
```

Comment:
I think isset is more appropriate here. At least that is what the get function is using